### PR TITLE
MAPL: add cray-mpich for MPI_STACK

### PR DESCRIPTION
**Pull request by AlexanderRichert-NOAA**
This PR adds cray-mpich to the possible MPI_STACK values in the MAPL recipe. Without it, compiling `mapl ^cray-mpich` fails.

### CODE REVIEW

**Comment by AlexanderRichert-NOAA** [comment]
@mathomp4 I'm not sure exactly what the MPI_STACK variable gets used for, so I'm making a guess as to the appropriate value. Let me know if that should be changed.

**Review by Hang-Lei-NOAA — APPROVED** [review]

**Review by climbfuji — APPROVED** [review]

File: var/spack/repos/builtin/packages/mapl/package.py [hunk 1]
```diff
@@ -349,6 +350,8 @@ def cmake_args(self):
             args.append(self.define("MPI_STACK", "mvapich"))
         elif self.spec.satisfies("^mpt"):
             args.append(self.define("MPI_STACK", "mpt"))
+        elif self.spec.satisfies("^cray-mpich"):
+            args.append(self.define("MPI_STACK", "cray-mpich"))
```
Comment (mathomp4) [inline var/spack/repos/builtin/packages/mapl/package.py]:
```suggestion
            args.append(self.define("MPI_STACK", "mpich"))
```

**Comment by mathomp4** [comment]
> @mathomp4 I'm not sure exactly what the MPI_STACK variable gets used for, so I'm making a guess as to the appropriate value. Let me know if that should be changed.

@AlexanderRichert-NOAA It's my attempt to tell our scripting what MPI stack is used so it knows what MPI environment variables to use. 

My guess is you might want to say cray-mpich → mpich since the [code I have in `ESMA_cmake`](https://github.com/GEOS-ESM/ESMA_cmake/blob/main/external_libraries/DetermineMPIStack.cmake) only knows a few stacks (the ones I can test). But, well, cray-mpich sounds...mpich-y.

Note: This is not technically a compile time issue, rather it is used by CMake to fill in scripts via `configure_file()` so that scripts know the MPI stack. I used to go off of "words in the path in a Baselibs install" but once spack appeared, well, that was useless and of course, CMake just knows "I have mpi" not "I have this flavor of mpi".

**Comment by mathomp4** [comment]
I guess on that end, can anyone here think of other MPI stacks we should put here? I don't have access to a Cray thus I've never tried it and I'm sure there might be others?

The other thing we could do, @AlexanderRichert-NOAA is you could provide the output of:
```cmake
  message(STATUS "MPI_STACK not specified. Attempting to autodetect MPI stack...")
  message(DEBUG "MPI_Fortran_LIBRARY_VERSION_STRING: ${MPI_Fortran_LIBRARY_VERSION_STRING}")

  string(REPLACE " " ";" MPI_Fortran_LIBRARY_VERSION_LIST ${MPI_Fortran_LIBRARY_VERSION_STRING})
  message(DEBUG "MPI_Fortran_LIBRARY_VERSION_LIST: ${MPI_Fortran_LIBRARY_VERSION_LIST}")
  list(GET MPI_Fortran_LIBRARY_VERSION_LIST 0 MPI_Fortran_LIBRARY_VERSION_FIRSTWORD)
  message(DEBUG "MPI_Fortran_LIBRARY_VERSION_FIRSTWORD: ${MPI_Fortran_LIBRARY_VERSION_FIRSTWORD}")
```

That is essentially what the code in ESMA_cmake is doing to try and figure things out. Then we could add Cray MPICH "for real".

**Comment by AlexanderRichert-NOAA** [comment]
That code doesn't give any output because MPI_Fortran_LIBRARY_VERSION_STRING is empty. It fails with a complaint about a missing argument for `string(REPLACE ...`

**Comment by mathomp4** [comment]
> ```cmake
> MPI_Fortran_LIBRARY_VERSION_STRING
> ```

Grah. I forgot I think you need a special flag before finding MPI:

```cmake
  set(MPI_DETERMINE_LIBRARY_VERSION TRUE)
  find_package(MPI)
```

**Comment by AlexanderRichert-NOAA** [comment]
That seems to be used. Maybe FindMPI.cmake isn't returning anything for it? I'll see what I can find.

**Comment by AlexanderRichert-NOAA** [comment]
Okay I think I more or less figured out the deal there. Here's what FindMPI.cmake gives for MPI_Fortran_LIBRARY_VERSION_STRING:
```console
  MPI VERSION : CRAY MPICH version 8.1.25.17 (ANL base 3.4a2)

  MPI BUILD INFO :

    Sun Feb 26 16:01 2023 (git hash aecd99f)
```

**Review by mathomp4 — APPROVED** [review]

**Comment by mathomp4** [comment]
> Okay I think I more or less figured out the deal there. Here's what FindMPI.cmake gives for MPI_Fortran_LIBRARY_VERSION_STRING:
> 
> ```
>   MPI VERSION : CRAY MPICH version 8.1.25.17 (ANL base 3.4a2)
> 
>   MPI BUILD INFO :
> 
>     Sun Feb 26 16:01 2023 (git hash aecd99f)
> ```

Well, I think from that it would pretty much return `mpich` and maybe even get the version right (though I don't use that anywhere yet). So, I think you got it right! Approved!

**Review by mathomp4 — APPROVED** [review]

**Review by alalazo — DISMISSED** [review]

**Comment by alalazo** [comment]
There's a conflict that needs resolution

**Comment by AlexanderRichert-NOAA** [comment]
These changes already incorporated in [a6ac78c](https://github.com/spack/spack/commit/a6ac78c7c6e0c740a0064b8673e16666710f228a)

**Pull request closed without merging**